docs/src/proposals/traced-transaction-construction.md

```diff
@@ -0,0 +1,247 @@
+# Traced transaction construction via RPC's `simulateTransaction`
+
+## Problem
+
+With more composition comes in the Solana's program ecosystem, it will be harder
+and harder for the higher-level consuming program to correctly construct
+lower-level program's instructions by precisely specifying each and every
+accessed accounts with the `read_only` or `read_write` flag.
+
+Also, there will be no way to signal the need of additional accounts for the
+arbitrarily customized behavior from the lower-level program to the higher-level
+one under the standardized instruction API like spl-token, even though we're
+about to explore the flexibility of allowing customzed programs conforming to a
+standard like the itoken.
+
+These preceding runtime limitations are applicable both to on-chain (instruction
+construction for CPI) or off-chain (transaction construction by end-user
+interfaces) boundaries likewise.
+
+Lastly, it's practically impossible to change those accessed account addresses
+once deployed while retaining compatibility, given the fact the consumers are
+expected to hard-code the account addresses for consumption, even if the
+deployed programs can be now upgradeable. This severely hampers ongoing
+consumed-program's implementation refinement in the areas of data migration and
+improved concurrency, by preventing transparent account switching and
+sharding with vectorized accounts, respectively.
+
+## Proposed Solution
+
+Extend the `simulateTransaction` RPC method to return more information like
+additional account addresses to read and/or write, collected from traced
+transaction execution in the simulation.
+
+Then, client signs the final transaction after appending these additional
+account addresses and submits it to the cluster via `sendTransaction`.
+
+For that end, introduce a new syscall to trace account accesses, so that
+programs can dynamically hint their _internal_ or _derivable_ account addresses
+implicitly, attaning the classic implementation-detail abstraction in the
+context Soalna's program ecosystem.
+
+In general, wallets should be able to manage these transaction finalization,
+agnostic to individual defi programs. Also, this extra simulation step results
+in no-op and harmless if all of the executed programs aren't aware of this new
+functionality.
+
+## Traced syscalls
+
+The runtime exposes new syscall to provide new account access mechanism which
+can be traced:
+
+```rust
+fn access_as_unsigned(address: &Pubkey, flags = enum { READ, WRITE}): 
```

Comment:
Considering this https://github.com/solana-labs/solana/pull/17901,

Maybe we should augment this by returning Enum<AccountInfo, AccountRef> with new flag ACCESS_AS_REF. so that we can avoid any serializing into program execution environment.

AccountRef can only be used for passing to child cpi.